### ISSUE TITLE
inconsistent prediction results using HinSage 

### ISSUE THREAD

**Issue by enjieli** [stellargraph/stellargraph#1900]
###  getting inconsistent prediction results after fitting a HinSage model 

I am using the HinSage-link-prediction demo [script](https://colab.research.google.com/github/stellargraph/stellargraph/blob/master/demos/link-prediction/hinsage-link-prediction.ipynb). 

After model fitting, I ran `model.predict(test_gen)` and got results

```
array([[3.9639022],
       [3.1461685],
       [3.5686603],
       ...,
       [4.0026994],
       [3.790706 ],
       [3.9910824]], dtype=float32)
```

ran `model.predict(test_gen)` again, this time got results

```
array([[4.011217 ],
       [3.3255208],
       [3.701041 ],
       ...,
       [4.098828 ],
       [3.6664836],
       [3.870615 ]], dtype=float32)
```

I have also set random seeds as 
```
import os 
import tensorflow as tf 
import random as python_random  

seed_value= 0
np.random.seed(123)
tf.random.set_seed(1234)
python_random.seed(123)
os.environ['PYTHONHASHSEED']="3"
```

I am using Stellagraph 1.2.1. 

Please advise how to fix this issue. Thank you!

**Comment by enjieli** [comment]
@huonw any suggestions?

**Comment by huonw** [comment]
Hi @enjieli thanks for waiting!

Unfortunately, I don't think HinSAGE is fully reproducible at the moment: https://github.com/stellargraph/stellargraph/pull/926 is a (very old) work-in-progress pull request that fixes this.

In any case, from your description I'm slightly unclear: are you setting the random seeds before each `predict` call, or just once at the start of the script? I think you should be able to slightly more reproducible behaviour using https://stellargraph.readthedocs.io/en/stable/api.html#module-stellargraph.random too:

```python
sg.random.set_seed(123)
tf.random.set_seed(123)
first = model.predict(test_gen)

sg.random.set_seed(123)
tf.random.set_seed(123)
second = model.predict(test_gen)
```

However, this won't be completely consistent 😦

**Comment by enjieli** [comment]
@huonw thanks for your response. I was setting the random seeds just once at the start of the script. 
Another quick question on HinSage, I have one node type that has 5000+  features, is it better to do feature dimension reduction first before building the graph? I have 4430 nodes and 2569835 edges in the graph;  each mode fit epoch takes a while to run. What's the best practice for my case? Thank you so so much!

**Comment by huonw** [comment]
You're right that large feature vectors may be suboptimal, because it does force copying a lot of data for every data point. HinSAGE samples a neighbourhood of each requested head node for training and prediction, and then has to collect the features of every node in the neighbourhood. The sizes of these neighbours can quickly become very large because the node types and edge types and number of samples multiply, rather than add.

That is to say, reducing feature size may improve speed significantly.

**Comment by enjieli** [comment]
@huonw thank you for your input! 

Are there ways to get node feature importance in the HinSage Link Prediction model? like the GCN `saliency_maps` .

**Comment by mikhail247** [comment]
I found that a possible workaround is to set batch_size to 1 in HinSAGENodeGenerator, which is used to call model.predict() method.  Other seed parameters need to be set as well.
The results become deterministic across different runs, however, this hurts prediction performance.
